### ISSUE TITLE
ProxQP on old Clang/GCC

### ISSUE THREAD

**Issue by costashatz** [Simple-Robotics/proxsuite#253]
Hello guys,

Thanks for the great work! I have been very satisfied by this solver so far.

I am having trouble getting it compiled in old compilers. Long story but I need to compile it with clang5. I am getting many errors related to `std::launder`, `std::aligned_alloc`, etc..

I am attaching a [Dockerfile](https://drive.google.com/file/d/1AeBVtS7nFMdSLiKyXSHEkFcJdp1WEqIi/view?usp=sharing) with the system that I need ProxQP to compile with. I hope this will help you to fix the compilation errors.

Any help will be appreciated.

Thanks in advance!

**Comment by jcarpent** [comment]
I've tried, but there are many places where the code needs to be changed and at some point, I think those compilers are not fully compatible with C++14 or C++17.
@costashatz Why you cannot select a more recent compiler? This point is not clear to me.

**Comment by costashatz** [comment]
Thanks a lot for the investigation so far.

So I managed to update to Clang 7.1.0 and gcc 7.3.1 (that's the most that I can go). More specifically:

`g++ (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)`

`clang version 7.1.0, Target: x86_64-unknown-linux-gnu`

I am still getting errors like the following:

```bash
In file included from proxsuite/examples/cpp/first_example_dense.cpp:7:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/proxqp/dense/dense.hpp:8:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/proxqp/dense/wrapper.hpp:10:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/proxqp/dense/solver.hpp:13:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/proxqp/dense/linesearch.hpp:9:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/proxqp/dense/model.hpp:11:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/proxqp/sparse/model.hpp:9:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/linalg/sparse/core.hpp:9:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/linalg/veg/memory/dynamic_stack.hpp:6:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/linalg/veg/internal/collection_algo.hpp:4:
In file included from /workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/linalg/veg/memory/alloc.hpp:10:
/workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/linalg/veg/memory/placement.hpp:63:35: error: no member named 'launder' in namespace 'std'
  const VEG_NOEXCEPT->T* { return VEG_LAUNDER(mem); }
                                  ^~~~~~~~~~~~~~~~
/workdir/proxsuite/build/install/lib/pkgconfig/../../include/proxsuite/linalg/veg/memory/placement.hpp:46:32: note: expanded from macro 'VEG_LAUNDER'
#define VEG_LAUNDER(p) (::std::launder(p))
                        ~~~~~~~^
```

I am using this command to compile the example: `clang++ -O3 -march=native -DNDEBUG -std=gnu++17 proxsuite/examples/cpp/first_example_dense.cpp -o first_example_dense $(pkg-config --cflags proxsuite)`

I have built `proxsuite` from source with `cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DBUILD_WITH_VECTORIZATION_SUPPORT=OFF`. Also, I get exactly the same errors with g++.

A Dockerfile to reproduce is:

```Dockerfile

FROM mottosso/mayabase-centos7

# Expose ninja, cmake3
RUN yum install -y http://repo.okay.com.mx/centos/7/x86_64/release/okay-release-1-1.noarch.rpm

# Build dependencies
RUN yum install -y centos-release-scl centos-release-scl-rh scl-utils
RUN yum install -y \
    ninja-build \
    cmake3 \
    libXt-devel \
    freeglut-devel \
    devtoolset-7

RUN ln -s /usr/bin/cmake3 /usr/bin/cmake

ENV CC=/opt/rh/devtoolset-7/root/usr/bin/gcc
ENV CXX=/opt/rh/devtoolset-7/root/usr/bin/g++
```

By the way (although it cannot solve my issue), if I use `devtoolset-11` and `clang11` everything works like a charm, **but I cannot use this version of the compiler**.

**Comment by costashatz** [comment]
Making the following patch:

```patch
diff --git a/include/proxsuite/linalg/veg/memory/alloc.hpp b/include/proxsuite/linalg/veg/memory/alloc.hpp
index 4c9056847..3ed961ad9 100644
--- a/include/proxsuite/linalg/veg/memory/alloc.hpp
+++ b/include/proxsuite/linalg/veg/memory/alloc.hpp
@@ -171,7 +171,7 @@ aligned_alloc(usize align, usize size) noexcept -> void*
   return alignment::detail::aligned_alloc(align, (size + mask) & ~mask);
 #endif
 #else
-#ifdef PROXSUITE_WITH_CPP_17
+#if defined(PROXSUITE_WITH_CPP_17) && defined(_LIBCPP_HAS_ALIGNED_ALLOC)
   return std::aligned_alloc(align, (size + mask) & ~mask);
 #else
   return alignment::detail::aligned_alloc(align, (size + mask) & ~mask);
diff --git a/include/proxsuite/linalg/veg/memory/placement.hpp b/include/proxsuite/linalg/veg/memory/placement.hpp
index 0ec1708cf..cb1afbc8a 100644
--- a/include/proxsuite/linalg/veg/memory/placement.hpp
+++ b/include/proxsuite/linalg/veg/memory/placement.hpp
@@ -41,7 +41,7 @@
 
 #if VEG_HAS_BUILTIN(__builtin_launder) || __GNUC__ >= 7
 #define VEG_LAUNDER(p) (__builtin_launder(p))
-#elif defined(VEG_WITH_CXX17_SUPPORT)
+#elif defined(VEG_WITH_CXX17_SUPPORT) && __GNUC__ >= 6
 #include <new>
 #define VEG_LAUNDER(p) (::std::launder(p))
 #else
 ```
 
 I was able to make ProxSuite compile and run using both g++ and clang. I am not sure if the patch is good but checking just for c++17 seems to not be enough.

**Comment by jcarpent** [comment]
@costashatz Could you open a PR? Your fix looks good to me.

**Comment by costashatz** [comment]
> @costashatz Could you open a PR? Your fix looks good to me.

Sure. Here it is: https://github.com/Simple-Robotics/proxsuite/pull/255

**Comment by costashatz** [comment]
Partly solved by #255.. clang5/gcc 4.8 are too old for ProxQP.

**Comment by costashatz** [comment]
Thanks a lot @jcarpent!
